lib/Service/Classification/ImportanceClassifier.php

```diff
@@ -0,0 +1,306 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OCA\Mail\Service\Classification;
+
+use Horde_Imap_Client;
+use OCA\Mail\Account;
+use OCA\Mail\Db\Classifier;
+use OCA\Mail\Db\Mailbox;
+use OCA\Mail\Db\MailboxMapper;
+use OCA\Mail\Db\Message;
+use OCA\Mail\Db\MessageMapper;
+use OCA\Mail\Service\Classification\FeatureExtraction\CompositeExtractor;
+use OCA\Mail\Support\PerformanceLogger;
+use OCA\Mail\Vendor\Phpml\Classification\NaiveBayes;
+use OCA\Mail\Vendor\Phpml\Estimator;
+use OCA\Mail\Vendor\Phpml\Metric\ClassificationReport;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\ILogger;
+use RuntimeException;
+use function array_column;
+use function array_combine;
+use function array_filter;
+use function array_map;
+use function array_slice;
+
+class ImportanceClassifier {
+
+	/**
+	 * Mailbox special uses to exclude from the training
+	 */
+	private const EXEMPT_FROM_TRAINING = [
+		Horde_Imap_Client::SPECIALUSE_ALL,
+		Horde_Imap_Client::SPECIALUSE_DRAFTS,
+		Horde_Imap_Client::SPECIALUSE_FLAGGED,
+		Horde_Imap_Client::SPECIALUSE_JUNK,
+		Horde_Imap_Client::SPECIALUSE_SENT,
+		Horde_Imap_Client::SPECIALUSE_TRASH,
+	];
+
+	/**
+	 * @var string label for data sets that are classified as important
+	 */
+	private const LABEL_IMPORTANT = 'i';
+
+	/**
+	 * @var string label for data sets that are classified as not important
+	 */
+	private const LABEL_NOT_IMPORTANT = 'ni';
+
+	/**
+	 * The maximum number of data sets to train the classifier with
+	 */
+	private const MAX_TRAINING_SET_SIZE = 1000;
+
+	/** @var MailboxMapper */
+	private $mailboxMapper;
+
+	/** @var MessageMapper */
+	private $messageMapper;
+
+	/** @var CompositeExtractor */
+	private $extractor;
+
+	/** @var PersistenceService */
+	private $persistenceService;
+
+	/** @var PerformanceLogger */
+	private $performanceLogger;
+
+	/** @var ILogger */
+	private $logger;
+
+	public function __construct(MailboxMapper $mailboxMapper,
+								MessageMapper $messageMapper,
+								CompositeExtractor $extractor,
+								PersistenceService $persistenceService,
+								PerformanceLogger $performanceLogger,
+								ILogger $logger) {
+		$this->mailboxMapper = $mailboxMapper;
+		$this->messageMapper = $messageMapper;
+		$this->extractor = $extractor;
+		$this->persistenceService = $persistenceService;
+		$this->performanceLogger = $performanceLogger;
+		$this->logger = $logger;
+	}
+
+	/**
+	 * Train an account's classifier of important messages
+	 *
+	 * Train a classifier based on a user's existing messages to be able to derive
+	 * importance markers for new incoming messages.
+	 *
+	 * To factor in (server-side) filtering into multiple mailboxes, the algorithm
```

Comment:
Not sure about and do not work for me :) I have a server side filter for GitHub notifications to move them out of the inbox because they are less important. If all non special mailboxes are scanned by default please make this configurable.

Reply:
What is your concern? That the app scans too many non-relevant mailboxes?

Reply:
That non-important messages are used to train the classifier.

Reply:
It should be a mix of both. Like so the classifier can learn what you think is important and what not. But it will indeed create a bias towards tagging as not-important if you have lots of these notifications. I'll have to think about it. Before we do any premature optimization we may also just try with your account and see how it performs :)